### PR TITLE
Capability statement enhancements/refactor

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description

- Large refactor of Capability generation
  - Removes DefaultCapabilities.json and AllCapabilities.json
  - Moves capability code into Core
- Adds ICapabilityStatementBuilder

## Related issues
Addresses [AB#70801](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/70801).

## Testing
- Unit tests
- E2E Tests

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/XmlFormatter/XmlFormatterConfiguration.cs [hunk 1]
```diff
@@ -39,15 +40,16 @@ public void PostConfigure(string name, MvcOptions options)
         {
             if (_featureConfiguration.SupportsXml)
             {
-                _configuredConformanceProvider.ConfigureOptionalCapabilities(statement => statement.Format = statement.Format.Concat(new[] { KnownContentTypes.XmlContentType }));
+                _configuredConformanceProvider
+                    .ConfigureOptionalCapabilities(statement => statement.Format.Add(KnownContentTypes.XmlContentType));
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/XmlFormatter/XmlFormatterConfiguration.cs]: This might not be needed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 2]
```diff
@@ -48,20 +49,23 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
...
             EnsureArg.IsNotNull(searchParameterTypeMap, nameof(searchParameterTypeMap));
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
             EnsureArg.IsNotNull(upsertResourceTvpGenerator, nameof(upsertResourceTvpGenerator));
             EnsureArg.IsNotNull(logger, nameof(logger));
             EnsureArg.IsNotNull(coreFeatures, nameof(coreFeatures));
 
             _configuration = configuration;
             _model = model;
             _searchParameterTypeMap = searchParameterTypeMap;
+            _modelInfoProvider = modelInfoProvider;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: Is this still being used?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                };
+                listedRestComponent.Resource.Add(resourceComponent);
+            }
+
+            action(resourceComponent);
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder TryAddRestInteraction(string resourceType, string interaction)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: nit: there is no "Try" here right? It will always add the interaction?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                    c.SearchParam.Add(searchParam);
+                }
+            });
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder AddDefaultResourceInteractions()
+        {
+            foreach (var resource in _modelInfoProvider.GetResourceTypeNames())
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: nit: string instead of var?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs (hunk 3 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: All the collections are HashSets so it's possible the call will be ignored. I'd also be ok with changing it to `AddRestInteraction`

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            {
+                if (string.Equals(resource, KnownResourceTypes.Parameters, StringComparison.Ordinal))
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: nit: maybe a bit of comment here will help people later? Why we are skipping certain types and what not?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                        Name = x.Name,
+                        Type = x.Type,
+                        Definition = x.Url,
+                        Documentation = x.Description,
+                    }));
+
+                    TryAddRestInteraction(resource, TypeRestfulInteraction.SearchType);
+                }
+
+                TryAddRestInteraction(resource, TypeRestfulInteraction.HistoryType);
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: Should these two be added in the DefaultResourceInteractions above? Doesn't seem to be related to search paramters?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs (hunk 3 above)
Comment (jackliums) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: Yeah after reading towards the end where JsonSerializer was created, I just realized that it might be hashset.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public bool Equals(T x, T y)
+        {
+            if (x == null || y == null)
+            {
+                return false;
+            }
+
+            return _propertiesToCompare.All(property => ReferenceEquals(property(x), property(y)) || property(x) == property(y));
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs]: nit: save the variable so it doesn't have to evaluate again?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            return this;
+        }
+
+        public ICapabilityStatementBuilder TryAddRestInteraction(string resourceType, string interaction)
+        {
+            EnsureArg.IsNotNullOrEmpty(resourceType, nameof(resourceType));
+            EnsureArg.IsNotNullOrEmpty(interaction, nameof(interaction));
+            EnsureArg.IsTrue(_modelInfoProvider.IsKnownResource(resourceType), nameof(resourceType), x => GenerateTypeErrorMessage(x, resourceType));
+
+            UpdateRestResourceComponent(resourceType, c =>
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: Is the capability statement built before the server starts? (e.g., does this need to be thread-safe?)

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/IProvideCapability.cs [hunk 9]
```diff
@@ -13,7 +13,7 @@ public interface IProvideCapability
         /// <summary>
         /// Allows this component to add capabilities into the specified ListedCapabilityStatement.
         /// </summary>
-        /// <param name="statement">The ListedCapabilityStatement to be added to.</param>
-        void Build(IListedCapabilityStatement statement);
+        /// <param name="builder">The ListedCapabilityStatement to be added to.</param>
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/IProvideCapability.cs]: nit: comment needs to be updated.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/ListedConformanceStatementExtensions.cs [hunk 10]
```diff
@@ -4,23 +4,19 @@
...
-        bool? ConditionalCreate { get; set; }
-
-        bool? ReadHistory { get; set; }
-
-        IList<string> SearchRevInclude { get; }
-
-        IList<string> SearchInclude { get; }
-
-        void AddResourceVersionPolicy(string policy);
+            return restComponents.Single(x => x.Mode == ListedCapabilityStatement.ServerMode);
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/ListedConformanceStatementExtensions.cs]: The `ListedCapabilityStatement.Rest` is already a hashset using `PropertyEqualityComparer` with Mode. Would that be utilized here when finding element. If not, can we utilize it?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public int GetHashCode(T obj)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs]: nit: obj could be null

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+            serializer.Serialize(writer, defaultOption);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override bool CanConvert(Type objectType)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs]: nit: set CanRead = false?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+{
+    internal class DefaultOptionHashSetJsonConverter : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            EnsureArg.IsNotNull(writer, nameof(writer));
+            EnsureArg.IsNotNull(value, nameof(value));
+            EnsureArg.IsNotNull(serializer, nameof(serializer));
+
+            dynamic defaultOption = ((dynamic)value).DefaultOption;
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs]: Instead of using dynamic, could define an interface like IDefaultOption or something which expose a property `object DefaultOption` then cast the object to that interface and access the property? It guards against potential renaming in the future and also no performance hit?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                    }
+
+                    if (defaultOption == item)
+                    {
+                        exists = true;
+                        break;
+                    }
+                }
+
+                if (!exists)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs]: Maybe a bit comment about why this logic exist would be nice. When would the default option won't exist and why the first item would be okay in that case?

Comment (jackliums) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs]: Sounds like even though the item is hashset, but it will only serialize one object? is that correct understanding?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/EnumLiteralJsonConverter.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+namespace Microsoft.Health.Fhir.Core.Features.Conformance.Serialization
+{
+    internal class EnumLiteralJsonConverter : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            EnsureArg.IsNotNull(writer, nameof(writer));
+            EnsureArg.IsNotNull(serializer, nameof(serializer));
+
+            if (value is Enum obj)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/EnumLiteralJsonConverter.cs]: FhirUtility already has a method called `GetLiteral`, which I think does this. It's in Hl7.Fhir.Support NuGet package which i think is version agnostic?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+                {
+                    _configurationUpdates.Clear();
+                    _sem.Release();
+                }
+            }
+
+            return _listedCapabilityStatement;
+        }
+
+        public void ConfigureOptionalCapabilities(Action<ListedCapabilityStatement> builderAction)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs]: Is this guaranteed to be called before GetCapabilityStatementAsync is called?

Comment (jackliums) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs]: I guess you do have guard against it. I wonder if we should consider making sure the capability statement is built as part of the warm up before the server starts taking requests?

File: src/Microsoft.Health.Fhir.Core/Models/CodingInfo.cs [hunk 17]
```diff
@@ -21,5 +22,10 @@ public CodingInfo(string system, string code)
         public string System { get; set; }
 
         public string Code { get; set; }
+
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Models/CodingInfo.cs]: Do we still need this? `Coding` seems to be in Hl7.Fhir.Support NuGet package now so that's version agnostic right?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs (hunk 6 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: That should be fine. History capabilities are tied to search and not the DataStore which is why they originated here.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs (hunk 8 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: We lock inside `SystemConformanceProvider`

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/ListedConformanceStatementExtensions.cs (hunk 10 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/ListedConformanceStatementExtensions.cs]: I don't think so? PropertyEqualityComparer is ensuring only 1 RestComponent of each Mode is added (Server/Client). This is explicitly fetching the Server rest component.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs (hunk 7 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs]: Not sure what you mean? This evaluates all specified properties on ObjA against the same properties on ObjB. The instances are passed in for comparison. `_propertiesToCompare` is an array of compiled functions (getters), so should also be fast.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs (hunk 14 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs]: Yes, even though our server can support multiple options, the capability statement only has one (in these cases). In a later task I want to use these sets to create a JSON Schema that defines all the valid options you can use to override

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/DefaultOptionHashSetJsonConverter.cs]: Also, perhaps the first item is not ok, perhaps it should throw an exception if the option is misconfigured.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs (hunk 16 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs]: Makes sense

File: src/Microsoft.Health.Fhir.Core/Models/CodingInfo.cs (hunk 17 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Models/CodingInfo.cs]: We don't need this class anymore, I think I'll defer this to a different PR

File: src/Microsoft.Health.Fhir.Core/Models/ModelInfoProvider.cs [hunk 18]
```diff
@@ -89,6 +90,12 @@ public static EvaluationContext GetEvaluationContext(ITypedElement element)
             return _modelInfoProvider.GetEvaluationContext(element);
         }
 
+        public static IStructureDefinitionSummaryProvider GetStructureDefinitionSummaryProvider()
+        {
+            EnsureProvider();
+            return _modelInfoProvider.GetStructureDefinitionSummaryProvider();
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Models/ModelInfoProvider.cs]: nit: could define a property and wrap around _modelInfoProvider

```
private static IModelInfoProvider
{
    get
    {
        EnsureProvider();

        return _modelInfoProvider;
    }
}
```

File: src/Microsoft.Health.Fhir.Shared.Core/VersionSpecificModelInfoProvider.cs [hunk 19]
```diff
@@ -59,5 +60,10 @@ public EvaluationContext GetEvaluationContext(ITypedElement element)
         {
             return new FhirEvaluationContext(element);
         }
+
+        public IStructureDefinitionSummaryProvider GetStructureDefinitionSummaryProvider()
+        {
+            return new PocoStructureDefinitionSummaryProvider();
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Shared.Core/VersionSpecificModelInfoProvider.cs]: Does this need new instance every time?

File: src/Microsoft.Health.Fhir.Shared.Core/VersionSpecificModelInfoProvider.cs [hunk 20]
```diff
@@ -42,7 +43,7 @@ public bool IsKnownCompartmentType(string compartmentType)
 
         public string[] GetResourceTypeNames()
         {
-            return Enum.GetNames(typeof(ResourceType));
+            return ModelInfo.SupportedResources.ToArray();
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Shared.Core/VersionSpecificModelInfoProvider.cs]: Seems like this could be cached.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs (hunk 7 above)
Comment (jackliums) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/PropertyEqualityComparer.cs]: `property(x)` and `property(y)` only need to be evaluated once instead of twice in the delegate.

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedCapabilityStatement.cs [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+using System.Collections.Generic;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Core.Features.Conformance.Models
+{
+    public class ListedCapabilityStatement
+    {
+        internal const string ServerMode = "server";
+        internal const string CapabilityStatement = "CapabilityStatement";
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedCapabilityStatement.cs]: This doesn't seem to be used?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedCapabilityStatement.cs [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+        public ListedCapabilityStatement()
+        {
+            Status = new DefaultOptionHashSet<string>("draft");
+            Kind = new DefaultOptionHashSet<string>("capability");
+            Rest = new HashSet<ListedRestComponent>(new PropertyEqualityComparer<ListedRestComponent>(x => x.Mode));
+            Format = new HashSet<string>();
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedCapabilityStatement.cs]: nit: StringComparer

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ConformanceBuilderTests.cs [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+            bool hasCreate = (bool)statement.Scalar($"{ResourceQuery("AuditEvent")}.interaction.where(code = '{TypeRestfulInteraction.Create}').exists()");
+            bool noUpdate = (bool)statement.Scalar($"{ResourceQuery("AuditEvent")}.interaction.where(code = '{TypeRestfulInteraction.Update}').exists()");
+            bool noDelete = (bool)statement.Scalar($"{ResourceQuery("AuditEvent")}.interaction.where(code = '{TypeRestfulInteraction.Delete}').exists()");
+
+            Assert.True(hasCreate);
+            Assert.False(noUpdate);
+            Assert.False(noDelete);
+        }
+
+        private static string ResourceQuery(string resource)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ConformanceBuilderTests.cs]: Should we also add some tests for Parameter and Binary since those also have some special logic. And one test for positive case like Pantient or something to make sure it has the correct capability built?

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ConformanceBuilderTests.cs]: Yes, added a test for Parameter, removed logic for Binary (I didn't find anything specific in the spec)

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedCapabilityStatement.cs [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace Microsoft.Health.Fhir.Core.Features.Conformance.Models
+{
+    public class ListedCapabilityStatement
+    {
+        internal const string ServerMode = "server";
+
+        public ListedCapabilityStatement()
+        {
+            Status = new DefaultOptionHashSet<string>("draft", StringComparer.Ordinal);
```
Comment (fmeira) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedCapabilityStatement.cs:19]: Hi @brendankowitz, how should the CapabilityStatement.Status be set to "active" if the model does not allow for changing it?
